File: Lucy/src/2025-8-W01/46_permutations.ts
function permute(nums: number[]): number[][] {
  const numsLen = nums.length;

  const results: number[][] = [];
  const visited: boolean[] = new Array(numsLen).fill(false);

  function backtrack(selected: number[]) {
    if (selected.length === numsLen) {
      results.push([...selected]);
      return;
    }

    for (let i = 0; i < numsLen; i++) {
      if (visited[i]) {
        continue;
      }

      visited[i] = true;
      selected.push(nums[i]);

      backtrack(selected);

      selected.pop();
      visited[i] = false;
    }
  }

  backtrack([]);

  return results;
}

File: Seori/46. Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        visited = [0] * n
        answer = []
        def backtracking(idx, num_list):
            if idx == n:
                answer.append(num_list[:])
                return
            
            # 순열은 순서에 상관이 없으므로 i의 인덱스를 0부터 세어준다.
            for i in range(len(nums)):
                if not visited[i]:
                    visited[i] = 1
                    num_list.append(nums[i])
                    backtracking(idx+1, num_list)
                    num_list.pop()
                    visited[i] = 0
            
        backtracking(0, [])
        return answer